lib/stack_master/template_compilers/sparkle_formation.rb

```diff
@@ -1,18 +1,52 @@
+require 'stack_master/sparkle_formation/compile_time/parameters_validator'
+require 'stack_master/sparkle_formation/compile_time/definitions_validator'
+require 'stack_master/sparkle_formation/compile_time/state_builder'
+
 module StackMaster::TemplateCompilers
   class SparkleFormation
+
+    CompileTime = StackMaster::SparkleFormation::CompileTime
+
     def self.require_dependencies
       require 'sparkle_formation'
       require 'stack_master/sparkle_formation/template_file'
     end
 
-    def self.compile(template_file_path, compiler_options = {})
-      if compiler_options["sparkle_path"]
-        ::SparkleFormation.sparkle_path = File.expand_path(compiler_options["sparkle_path"])
+    def self.compile(template_file_path, parameters, compiler_options = {})
+      if compiler_options['sparkle_path']
+        ::SparkleFormation.sparkle_path = File.expand_path(compiler_options['sparkle_path'])
       else
         ::SparkleFormation.sparkle_path = File.dirname(template_file_path)
       end
+      sparkle_template = ::SparkleFormation.compile(template_file_path, :sparkle)
+      definitions = sparkle_template.parameters
+      validate_definitions(definitions)
+      validate_parameters(definitions, parameters)
+
+      sparkle_template.compile_time_parameter_setter do
+        sparkle_template.compile_state = create_state(definitions, parameters)
+        remove_compile_parameters(definitions, parameters)
```

Comment:
Not keen on how this works... relying on parameters to this method being mutated so that SM doesn't use compile time params as CF params. I wonder if it would be better to define compile time params in `stack_master.yml` under the stack definition. That way we're not mixing the 2 types of parameters. Thoughts?

```
my-stack:
  template: body.json
  compiler_options:
    parameters:
      private_subnet_cidrs:
        - 10.0.0.0/32:ap-southeast-2
        - 10.0.0.2/32:ap-southeast-1
```

Reply:
You could. @patrobinson wanted to be able to also get other stacks outputs and feed them in as compile time parameters. Would it work with that?

Reply:
When you use compile time parameters in sparkle it automatically generates the parameters as stack outputs. I think they wanted you to be able to feed them as inputs to other stacks. So would it follow how people expect sparkle to work?

Reply:
No it wouldn't work in that case.

It looks like the output added is just a JSON dump of the `CompiledState`, which I assume is for debugging purposes or being able to recompile a version of the template by knowing the compile time params used. If you wanted to feed anything meaningful into another stack you'd have to manually create an output.

Reply:
btw if you care to have a look at how sparkle uses these things in their `sfn` utilty look no further than [here](https://github.com/sparkleformation/sfn/blob/develop/lib/sfn/command_module/template.rb)

Reply:
@stevehodgkiss Our intention is to keep it as close to how SparkleFormation works, so we dont have to create and maintain additional documentation.

Reply:
You are maintaining the API to SparkleFormation, that doesn't prevent splitting out how parameters are handled from StackMaster's point of view.

Reply:
The mutation did feel icky when we were doing it, but we were optimising to keep the behaviour as close to SparkleFormation as possible, while leaving it open for other compilers like cfndsl.

Also the `stack_master.yml` feels like  the wrong place to implement this. These are compile time paramters and not compiler options. So it feels like the parameters yaml might be a better place to flag some parameters as compile time parameters.

Reply:
I agree with @shivaman `stack_master.yml` also feels like the wrong place to me. Maybe something in the parameters.yml files like this

```
stream_name: cloudwatch-logs
shards: 1
   compile_time:
      private_subnet_cidrs:
        - 10.0.0.0/32:ap-southeast-2
        - 10.0.0.2/32:ap-southeast-1
```

I'm also thinking of how I would use them. i.e. create n-subnets in one stack. Output the subnet identifiers as a comma delimited string to be used as compile time parameters for the downstream stack.

Reply:
Sounds reasonable @gstib. That or a separate dir/file. Thoughts @patrobinson?

Reply:
I'm happy with that suggestion